### ISSUE TITLE
[Bug] Buffer overrun in BF_ofb64_encrypt

### ISSUE THREAD

**Issue by yiyuaner** [openssl/openssl#18751]
In the file `crypto/bf/bf_ofb64.c`, the function `BF_ofb64_encrypt` has the following [code](https://github.com/openssl/openssl/blob/2837b19fcba4bf4ff2ecdc8435c650bf18c27552/crypto/bf/bf_ofb64.c#L24):

```
void BF_ofb64_encrypt(const unsigned char *in, unsigned char *out,
                      long length, const BF_KEY *schedule,
                      unsigned char *ivec, int *num) {
    register int n = *num;
    unsigned char d[8];
    while (l--) {
        ......
        *(out++) = *(in++) ^ d[n];
        n = (n + 1) & 0x07;
    }
}
```
The buffer `d` has a size of `8` bytes and inside the loop it is accessed with `d[n]`. However, `n=*num` may be a negative value. For example, one of the caller for `BF_ofb64_encrypt` is `bf_ofb_cipher` and has the following [code](https://github.com/openssl/openssl/blob/2837b19fcba4bf4ff2ecdc8435c650bf18c27552/include/crypto/evp.h#L371): 
```
int num = EVP_CIPHER_CTX_get_num(ctx);
BF_ofb64_encrypt(in, out, (long)EVP_MAXCHUNK, &EVP_C_DATA(kstruct,ctx)->ksched, ctx->iv, &num);
```
The return value for `num = EVP_CIPHER_CTX_get_num(ctx)` may be `-1` (see the code [here](https://github.com/openssl/openssl/blob/2837b19fcba4bf4ff2ecdc8435c650bf18c27552/crypto/evp/evp_lib.c#L617)) and is not checked before passing into the function `BF_ofb64_encrypt`.

**Comment by paulidale** [comment]
For BF (and the other ciphers in the default provider), this won't be an issue since the param is implemented.  For a third party provider, it could be.

**Comment by slontis** [comment]
For the legacy paths using engines (the e_ files).. it could only return -1 if ctx->cipher is NULL.

**Comment by slontis** [comment]
Its not likely to be a problem in the providers since they use an internal num inside prov_cipher_ctx_st

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev barring further input
